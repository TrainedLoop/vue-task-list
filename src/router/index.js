import { createWebHistory, createRouter } from 'vue-router';
import Hello from '@/views/ApiTest.vue';
import Home from '@/views/Home.vue';
const routes = [
  {
    path: '/',
    name: 'Home',
    component: Home,
  },
  {
    path: '/hello',
    name: 'hello',
    component: Hello,
  },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
