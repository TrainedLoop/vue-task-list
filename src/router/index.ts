import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';
import PublicLayout from '../layouts/PublicLayout.vue';
import LoginView from '@/views/LoginView.vue';
import RegisterView from '@/views/RegisterView.vue';

const routes: Array<RouteRecordRaw> = [
  {
    path: '/login',
    alias: ['/'],
    component: LoginView,
    meta: { layout: PublicLayout, transition: 'slide-left' },
  },
  {
    path: '/register',
    component: RegisterView,
    meta: { layout: PublicLayout, transition: 'slide-left' },
  },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
