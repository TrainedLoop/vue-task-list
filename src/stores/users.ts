import { defineStore } from 'pinia';

interface IUserStore {
  name?: string;
  email?: string;
  auth: boolean;
  loginLoading: boolean;
  loginError?: string;
}

interface ILogin {
  login: string;
  password: string;
}
export const useUserStore = defineStore('user', {
  state: (): IUserStore => ({
    name: undefined,
    email: undefined,
    auth: false,
    loginLoading: false,
    loginError: undefined,
  }),
  getters: {
    getUsers(state) {
      return state;
    },
  },
  actions: {
    async authUser({ login, password }: ILogin) {
      try {
        this.loginLoading = true;
      } catch (error) {
        alert(error);
        console.log(error);
      } finally {
        this.loginLoading = false;
      }
    },
  },
});
