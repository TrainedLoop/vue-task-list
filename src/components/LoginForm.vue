<script setup lang="ts">
import { ref } from 'vue';
import router from '@/router';
import { useUserStore } from '@/stores/users';
import { storeToRefs } from 'pinia';
import { computed } from '@vue/reactivity';
import GenericButton from './GenericButton.vue';
import GenericInput from './GenericInput.vue';
import GenericCard from './GenericCard.vue';

const store = useUserStore()
const { loginLoading, loginError } = storeToRefs(store);
const { authUser } = store

const login = ref("")
const password = ref("")

const buttonText = computed(() => {
    return loginLoading.value ? "Entrando..." : "Entrar"
})

const handleSignClick = async () => {
    if (loginLoading.value === true)
        return;
    await authUser({ login: login.value, password: password.value })
}
const handleRegisterClick = () => {
    router.push('/register')
}
</script>

<template>
    <GenericCard>
        <form class="login-form">
            <div class="login-title">
                <h1>Task List App</h1>
            </div>
            <div class="input-group">
                <GenericInput name="Login" v-model="login" />
                <GenericInput name="Senha" type="password" :errorMessage="loginError" v-model="password" />
            </div>
            <GenericButton class="submit-login" :text="buttonText" type="success" :onClick="handleSignClick" />
            <GenericButton class="register" text="Cadastrar" type="default" :onClick="handleRegisterClick" />
        </form>
    </GenericCard>
</template>

<style lang="scss" scoped>
.login-form {
    display: flex;
    width: 400px;
    flex-direction: column;
    justify-content: space-evenly;

    .login-title {
        display: flex;
        justify-content: center;
        color: $color-neutral-dark-1;
    }

    .input:first-child {
        margin-bottom: 2rem;
    }

    .submit-login {
        margin-top: 2rem;

    }
}
</style>