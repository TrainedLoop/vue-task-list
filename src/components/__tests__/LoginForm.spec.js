import { mount, shallowMount, config } from '@vue/test-utils';
import { nextTick } from 'vue';
import LoginForm from '../LoginForm.vue';
import { createTestingPinia } from '@pinia/testing';
import { useUserStore } from '@/stores/users';
import router from '@/router';

jest.mock('../../router', () => ({ push: jest.fn() }));

describe('LoginForm.vue Tests', () => {
  let wrapper = null;
  let store = null;
  beforeEach(() => {
    config.global.renderStubDefaultSlot = true;
    wrapper = mount(LoginForm, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              user: {
                loginLoading: false,
                loginError: undefined,
              },
            },
          }),
        ],
      },
    });
    store = useUserStore();
  });
  afterEach(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });
  afterAll(() => {
    config.global.renderStubDefaultSlot = false;
  });

  it('should renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should store has correct initial Value', () => {
    expect(store.loginLoading).toStrictEqual(false);
    expect(store.loginError).toStrictEqual(undefined);
  });

  it('should render two generic inputs', () => {
    const input = wrapper.findAllComponents({ name: 'GenericInput' });
    expect(input).toHaveLength(2);
  });

  it('should first input has correct props', () => {
    const input = wrapper.findAllComponents({ name: 'GenericInput' }).at(0);
    expect(input.props()).toStrictEqual({
      errorMessage: undefined,
      modelValue: '',
      name: 'Login',
    });
  });

  it('should second input is typed as password', () => {
    const input = wrapper.findAllComponents({ name: 'GenericInput' }).at(1);
    expect(input.attributes('type')).toBe('password');
  });

  it('should input button has correct props', () => {
    const input = wrapper.findAllComponents({ name: 'GenericInput' }).at(1);
    expect(input.props()).toStrictEqual({
      errorMessage: undefined,
      modelValue: '',
      name: 'Senha',
    });
  });

  it('should render two generic buttons', () => {
    const button = wrapper.findAllComponents({ name: 'GenericButton' });
    expect(button).toHaveLength(2);
  });

  it('should first button has correct props', () => {
    const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(0);
    expect(button.props()).toStrictEqual({
      onClick: expect.any(Function),
      text: 'Entrar',
      type: 'success',
    });
  });
  describe('if state loginLoading is true', () => {
    it('should first button has correct text prop', async () => {
      store.loginLoading = true;
      await nextTick();
      const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(0);
      expect(button.props('text')).toBe('Entrando...');
    });
  });
  describe('if first button click callback is called', () => {
    it('should call authUser from store 1 time', async () => {
      const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(0);
      await button.props().onClick();
      expect(store.authUser).toBeCalledTimes(1);
    });

    it('should call authUser from store with correct params', async () => {
      const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(0);
      await button.props().onClick();
      expect(store.authUser).toBeCalledWith({ login: '', password: '' });
    });

    describe('if state loginLoading is true', () => {
      it('should not call authUser from store', async () => {
        store.loginLoading = true;
        await nextTick();
        const button = wrapper
          .findAllComponents({ name: 'GenericButton' })
          .at(0);
        await button.props().onClick();
        expect(store.authUser).not.toBeCalled();
      });
    });
  });

  it('should second button has correct props', () => {
    const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(1);
    expect(button.props()).toStrictEqual({
      onClick: expect.any(Function),
      text: 'Cadastrar',
      type: 'default',
    });
  });

  describe('if second button click callback is called', () => {
    it('should call router.push 1 time', async () => {
      const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(1);
      await button.props().onClick();
      expect(router.push).toBeCalledTimes(1);
    });

    it('should call router.push with correct params', async () => {
      const button = wrapper.findAllComponents({ name: 'GenericButton' }).at(1);
      await button.props().onClick();
      expect(router.push).toBeCalledWith('/register');
    });
  });
});
