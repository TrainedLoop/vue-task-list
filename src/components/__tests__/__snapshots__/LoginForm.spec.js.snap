// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm.vue Tests should renders component correctly 1`] = `
<div
  class="card"
>
  
  <form
    class="login-form"
  >
    <div
      class="login-title"
    >
      <h1>
        Task List App
      </h1>
    </div>
    <div
      class="input-group"
    >
      <div
        class="input input--empty"
      >
        <input
          aria-labelledby="placeholder-Login"
          autocomplete="off"
          id="Login"
          name="Login"
        />
        <label
          for="Login"
          id="placeholder-Login"
        >
          <span
            class="label-text"
          >
            Login
          </span>
        </label>
        <div
          class="error"
        >
          <small />
        </div>
      </div>
      <div
        class="input input--empty"
        type="password"
      >
        <input
          aria-labelledby="placeholder-Senha"
          autocomplete="off"
          id="Senha"
          name="Senha"
          type="password"
        />
        <label
          for="Senha"
          id="placeholder-Senha"
        >
          <span
            class="label-text"
          >
            Senha
          </span>
        </label>
        <div
          class="error"
        >
          <small />
        </div>
      </div>
    </div>
    <div
      class="generic-button generic-button--success submit-login"
    >
      <button
        class="submit-login"
        type="button"
      >
        Entrar
      </button>
    </div>
    <div
      class="generic-button register"
    >
      <button
        class="register"
        type="button"
      >
        Cadastrar
      </button>
    </div>
  </form>
  
</div>
`;
